File: clj/src/advent_2017/day1.clj
(ns advent-2017.day1)

(defn specialReduce [coll]
  (let [indexedColl (zipmap (range (count coll)) coll) len (count coll)]
    (reduce (fn [acc e]
      (let [idx (if (= (key e) (- len 1)) 0 (+ 1 (key e)))]
        (if (= (nth coll idx) (val e)) (+ acc (val e)) acc)))
        0 indexedColl)))

(defn captcha [num]
  (specialReduce
    (map #(Integer. %)
      (clojure.string/split (str num) #""))))

File: js/lib/day1.js
const captcha = (num) => {
  const arr = num.split('').map((e) => parseInt(e))
  return arr.reduce((a, e, i) => {
    const idx = i + 1 === arr.length ? 0 : i + 1
    if (arr[idx] === e) {
      return a + e
    }
    return a
  }, 0)
}

module.exports = captcha
